From 1838ba96f8311cfcc4213e72641c9a3123464fc7 Mon Sep 17 00:00:00 2001
From: Stephane Desneux <stephane.desneux@iot.bzh>
Date: Tue, 16 Jul 2019 17:23:50 +0000
Subject: [PATCH] aglsetup: add --topic option to tag multiple builds with same
 ID

This patch adds the option -t|--topic to aglsetup (default: empty)

The topic value is propagated to aglsetup.manifest then used by
distro-build-manifest to put it in DIST_BUILD_TOPIC variable
in all build manifests (deploy dir, target image, sdk)

Using the --topic option to specify a unique identifier will help
to provide a common ID for all builds in the same "family" even
if features and layers differ. For this purpose, a topic could be:
* a gerrit review id + a patchset number
* a project name (git repo) + a Change-Id
* a random UUID
* ...

Bug-AGL: SPEC-2646

Change-Id: I0b68dfa297509dac07e9b2942948631cfc13c319
Signed-off-by: Stephane Desneux <stephane.desneux@iot.bzh>
---
 scripts/.aglsetup_genconfig.bash     | 17 ++++++++++++++---
 scripts/distro-manifest-generator.sh |  9 ++++++---
 2 files changed, 20 insertions(+), 6 deletions(-)

diff --git a/scripts/.aglsetup_genconfig.bash b/scripts/.aglsetup_genconfig.bash
index 6e091e2..603e25e 100755
--- a/scripts/.aglsetup_genconfig.bash
+++ b/scripts/.aglsetup_genconfig.bash
@@ -4,7 +4,7 @@
 #
 # The MIT License (MIT)
 #
-# Copyright (c) 2016 Stéphane Desneux <sdx@iot.bzh>
+# Copyright (c) 2016-2019 Stéphane Desneux <sdx@iot.bzh>
 #           (c) 2016 Jan-Simon Möller <jsmoeller@linuxfoundation.org>
 #
 # Permission is hereby granted, free of charge, to any person obtaining a copy
@@ -31,7 +31,7 @@
 # turn execute (source) generated instructions back in the parent shell,
 # whether it's bash, zsh, or any other supported shell
 
-VERSION=1.1.0
+VERSION=1.1.1
 DEFAULT_MACHINE=qemux86-64
 DEFAULT_BUILDDIR=./build
 VERBOSE=0
@@ -161,6 +161,12 @@ Options:
           'timestamp' : Use a generated time stamp (UTC).
           'value:<revision>' : Use <revision> explicitly.
           'none' : Do nothing.
+   -t|--topic <value>
+      Specify an optional topic for this setup.
+      If specified, the topic will be propagated in build manifests:
+         - in deployment dir: tmp/deploy/images/*/build-info
+         - in target image: /etc/platform-info/build
+         - in SDK: tmp/deploy/sdk/*.build-info
    -v|--verbose
       verbose mode
       default: false
@@ -293,7 +299,7 @@ function find_feature_dependency() {
 
 GLOBAL_ARGS=( "$@" )
 debug "Parsing arguments: $@"
-TEMP=$(getopt -o m:b:r:s:fvdh --long machine:,builddir:,rpm-revision:,script:,force,verbose,debug,help -n $SCRIPT -- "$@")
+TEMP=$(getopt -o m:b:r:t:s:fvdh --long machine:,builddir:,rpm-revision:,topic:,script:,force,verbose,debug,help -n $SCRIPT -- "$@")
 [[ $? != 0 ]] && { usage; exit 1; }
 eval set -- "$TEMP"
 
@@ -305,6 +311,7 @@ BUILDDIR=$DEFAULT_BUILDDIR
 SETUPSCRIPT=
 FORCE=
 RPMREVISION=
+TOPIC=
 SETUP_MANIFEST=aglsetup.manifest
 
 while true; do
@@ -314,6 +321,7 @@ while true; do
 		-s|--setupscript)  SETUPSCRIPT=$2; shift 2;;
 		-f|--force)        FORCE=1; shift;;
 		-r|--rpm-revision) RPMREVISION=$2; shift 2;;
+		-t|--topic)        TOPIC=$2; shift 2;;
 		-v|--verbose)      VERBOSE=1; shift;;
 		-d|--debug)        VERBOSE=1; DEBUG=1; shift;;
 		-h|--help)         HELP=1; shift;;
@@ -510,6 +518,9 @@ DIST_METADIR="$METADIR"
 # timestamp
 DIST_SETUP_TS="$(date -u +%Y%m%d_%H%M%S_%Z)"
 
+# topic
+DIST_SETUP_TOPIC="$TOPIC"
+
 # ------------ end of $SCRIPT fragment --------
 EOF
 info "OK"
diff --git a/scripts/distro-manifest-generator.sh b/scripts/distro-manifest-generator.sh
index c7e8cc3..0db3e13 100755
--- a/scripts/distro-manifest-generator.sh
+++ b/scripts/distro-manifest-generator.sh
@@ -194,6 +194,9 @@ function _getgitmanifest() {
 	# Manifest build timestamp
 	DIST_BUILD_TS="$timestamp"
 
+	# build topic from setup topic
+	DIST_BUILD_TOPIC="${DIST_SETUP_TOPIC}"
+
 	# what to retain from setup manifest?
 	# to generate the full list: cat setup.manifest  | grep = | cut -f1 -d"=" | awk '{printf("%s ",$1);}'
 	declare -A SETUP_VARS
@@ -203,9 +206,9 @@ function _getgitmanifest() {
 
 	# extra vars not coming from setup.manifest but generated here
 	declare -A EXTRA_VARS
-	EXTRA_VARS[deploy]="DIST_SETUP_MANIFEST DIST_BUILD_TS DIST_LAYERS DIST_LAYERS_MD5 DIST_BUILD_HASH DIST_BUILD_ID"
-	EXTRA_VARS[target]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS"
-	EXTRA_VARS[sdk]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS"
+	EXTRA_VARS[deploy]="DIST_SETUP_MANIFEST DIST_BUILD_TS DIST_LAYERS DIST_LAYERS_MD5 DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TOPIC"
+	EXTRA_VARS[target]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS DIST_BUILD_TOPIC"
+	EXTRA_VARS[sdk]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS DIST_BUILD_TOPIC"
 
 	# BITBAKE_VARS may be defined from external file to source (--source arg)
 	# this is used to dump extra vars from inside bitbake recipe
-- 
2.1.4

